src/com/google/enterprise/adaptor/database/DatabaseAdaptor.java

```diff
@@ -699,17 +689,38 @@ public static void main(String[] args) {
     AbstractAdaptor.main(new DatabaseAdaptor(), args);
   }
   
-  private Acl getAcl(Connection conn, String uniqueId) throws SQLException {
-    try (PreparedStatement stmt = getAclFromDb(conn, uniqueId);
-        ResultSet rs = stmt.executeQuery()) {
+  private Acl getAcl(Connection conn, ResultSet rs, String uniqueId)
+      throws SQLException {
+    log.log(Level.FINER, "about to get acl: {0}", uniqueId);
+    if (aclSql != null || rs == null) {
+      String sql = (aclSql == null) ? singleDocContentSql : aclSql;
+      try (PreparedStatement st = conn.prepareStatement(sql)) {
+        // aclSql will only be null when called from isUserAuthorized.
+        if (aclSql == null) {
+          uniqueKey.setContentSqlValues(st, uniqueId);
+          return processAclQuery(st, null);
+        } else {
+          uniqueKey.setAclSqlValues(st, uniqueId);
+          return processAclQuery(st, Acl.EMPTY);
+        }
+      }
+    } else {
+      // Generate ACL if it came as part of result to singleDocContentSql.
+      return buildAcl(rs, aclPrincipalDelimiter, aclNamespace,
+          ResultSetNext.PROCESS_ONE_ROW, null);
+    }
+  }
+
+  private Acl processAclQuery(PreparedStatement st, Acl defaultAcl)
+      throws SQLException {
+    try (ResultSet resSet = st.executeQuery()) {
       log.finer("got acl");
-      boolean hasResult = rs.next();
+      boolean hasResult = resSet.next();
```

Comment:
Why s/rs/resSet/? It's a non-standard abbreviation, and rs in a database connector seems perfectly clear.

Reply:
Unintended changes in this one, and carried over from trying to make it two nested try-with-resources blocks.